File: Pandacap.HighLevel/JsonLdExpansionService.cs
using JsonLD.Core;
using Newtonsoft.Json.Linq;

namespace Pandacap.HighLevel
{
    public class JsonLdExpansionService
    {
        private readonly Dictionary<string, RemoteDocument> _contexts = [];

        private class CustomDocumentLoader(JsonLdExpansionService service) : DocumentLoader
        {
            public override RemoteDocument LoadDocument(string url)
            {
                if (service._contexts.TryGetValue(url, out RemoteDocument? cached))
                    return cached;

                var fetched = base.LoadDocument(url);
                service._contexts.Add(url, fetched);
                return fetched;
            }
        }

        [System.Diagnostics.CodeAnalysis.SuppressMessage("Performance", "CA1822:Mark members as static", Justification = "Used in dependency injection")]
        public JToken Expand(JObject jObject)
        {
            try
            {
                return JsonLdProcessor.Expand(jObject).Single();
            }
            catch (JsonLdError ex) when (ex.GetType() == JsonLdError.Error.RecursiveContextInclusion || ex.GetType() == JsonLdError.Error.LoadingRemoteContextFailed)
            {
                // Remove everything except ActivityStreams from the context
                jObject["@context"] = new JArray(new JValue("https://www.w3.org/ns/activitystreams"));

                // Retry
                return JsonLdProcessor.Expand(jObject, new JsonLdOptions
                {
                    documentLoader = new CustomDocumentLoader(this)
                }).Single();
            }
        }
    }
}

File: Pandacap.LowLevel/ActivityPub/Communication/JsonLdExpansionService.fs
namespace Pandacap.ActivityPub.Communication

open System
open Microsoft.Extensions.Caching.Memory
open Newtonsoft.Json.Linq
open JsonLD.Core

type internal CustomDocumentLoader(cache: IMemoryCache) =
    inherit DocumentLoader()

    override _.LoadDocument (url: string): RemoteDocument = 
        let key = $"246d1dc3-a59e-49cc-a555-46d35635ac2d:{url}"

        match cache.TryGetValue(key) with
        | true, (:? RemoteDocument as cached) ->
            cached
        | _ ->
            cache.Set(
                key,
                base.LoadDocument(url),
                DateTimeOffset.UtcNow.AddHours(1))

type JsonLdExpansionService(cache: IMemoryCache) =
    member _.Expand(jObject: JObject) =
        let options = new JsonLdOptions(
            documentLoader = new CustomDocumentLoader(cache))

        try
            JsonLdProcessor.Expand(jObject, options) |> Seq.head
        with :? JsonLdError ->
            jObject["@context"] <- new JArray(
                new JValue("https://www.w3.org/ns/activitystreams"),
                new JValue("https://w3id.org/security/v1"))
            JsonLdProcessor.Expand(jObject, options) |> Seq.head
